drivers/rtc/rtc_nxp_irtc.c

```diff
@@ -0,0 +1,454 @@
+	}
+
+	if ((mask & RTC_ALARM_TIME_MASK_YEAR) == RTC_ALARM_TIME_MASK_YEAR) {
+		irtc_reg->ALM_YEARMON |=
+			RTC_ALM_YEARMON_ALM_YEAR(timeptr->tm_year - RTC_NXP_IRTC_YEAR_OFFSET);
+	}
+
+	/* Clearing out the ALARM Flag Field then setting the correct value */
+	irtc_reg->CTRL &= ~(0xC);
+	switch (mask) {
+	case 0x0F: {
+		irtc_reg->CTRL |= RTC_CTRL_ALM_MATCH(0x4);
+	} break;
+	case 0x1F: {
+		irtc_reg->CTRL |= RTC_CTRL_ALM_MATCH(0x8);
+	} break;
+	case 0x3F: {
+		irtc_reg->CTRL |= RTC_CTRL_ALM_MATCH(0xC);
+	} break;
+	default: {
+		irtc_reg->CTRL |= RTC_CTRL_ALM_MATCH(0x0);
+	}
+	}
+
+	/* Enabling Alarm Interrupts */
+	irtc_reg->IER |= 0x4;
+	data->alarm_mask = mask;
+
+	irq_unlock(key);
+
+	return 0;
+}
+
+static int nxp_irtc_alarm_get_time(const struct device *dev, uint16_t id, uint16_t *mask,
+				   struct rtc_time *timeptr)
+{
+	const struct nxp_irtc_config *config = dev->config;
+	struct nxp_irtc_data *data = dev->data;
+	RTC_Type *irtc_reg = config->base;
+	uint16_t curr_alarm_mask = data->alarm_mask;
+	uint16_t return_mask = 0;
+
+	if (id != 0 || !timeptr) {
+		return -EINVAL;
+	}
+
+	if ((curr_alarm_mask & RTC_ALARM_TIME_MASK_SECOND) == RTC_ALARM_TIME_MASK_SECOND) {
+		timeptr->tm_sec = ((irtc_reg->ALM_SECONDS & RTC_ALM_SECONDS_ALM_SEC_MASK) >>
+				   RTC_ALM_SECONDS_ALM_SEC_SHIFT);
+		return_mask |= RTC_ALARM_TIME_MASK_SECOND;
+	}
+
+	if ((curr_alarm_mask & RTC_ALARM_TIME_MASK_MINUTE) == RTC_ALARM_TIME_MASK_MINUTE) {
+		timeptr->tm_min = ((irtc_reg->ALM_HOURMIN & RTC_ALM_HOURMIN_ALM_MIN_MASK) >>
+				   RTC_ALM_HOURMIN_ALM_MIN_SHIFT);
+		return_mask |= RTC_ALARM_TIME_MASK_MINUTE;
+	}
+
+	if ((curr_alarm_mask & RTC_ALARM_TIME_MASK_HOUR) == RTC_ALARM_TIME_MASK_HOUR) {
+		timeptr->tm_hour = ((irtc_reg->ALM_HOURMIN & RTC_ALM_HOURMIN_ALM_HOUR_MASK) >>
+				    RTC_ALM_HOURMIN_ALM_HOUR_SHIFT);
+		return_mask |= RTC_ALARM_TIME_MASK_HOUR;
+	}
+
+	if ((curr_alarm_mask & RTC_ALARM_TIME_MASK_MONTHDAY) == RTC_ALARM_TIME_MASK_MONTHDAY) {
+		timeptr->tm_mday = ((irtc_reg->ALM_DAYS & RTC_ALM_DAYS_ALM_DAY_MASK) >>
+				    RTC_ALM_DAYS_ALM_DAY_SHIFT);
+		return_mask |= RTC_ALARM_TIME_MASK_MONTHDAY;
+	}
+
+	if ((curr_alarm_mask & RTC_ALARM_TIME_MASK_MONTH) == RTC_ALARM_TIME_MASK_MONTH) {
+		timeptr->tm_mon = ((irtc_reg->ALM_YEARMON & RTC_ALM_YEARMON_ALM_MON_MASK) >>
+				   RTC_ALM_YEARMON_ALM_MON_SHIFT) - 1;
+		return_mask |= RTC_ALARM_TIME_MASK_MONTH;
+	}
+
+	if ((curr_alarm_mask & RTC_ALARM_TIME_MASK_YEAR) == RTC_ALARM_TIME_MASK_YEAR) {
+		timeptr->tm_year =
+			(int8_t)((irtc_reg->ALM_YEARMON & RTC_ALM_YEARMON_ALM_YEAR_MASK) >>
+				 RTC_ALM_YEARMON_ALM_YEAR_SHIFT) + RTC_NXP_IRTC_YEAR_OFFSET;
+		return_mask |= RTC_ALARM_TIME_MASK_YEAR;
+	}
+
+	*mask = return_mask;
+
+	return 0;
+}
+
+static int nxp_irtc_alarm_is_pending(const struct device *dev, uint16_t id)
+{
+	struct nxp_irtc_data *data = dev->data;
+	int ret = data->is_alarm_pending;
+
+	if (id != 0) {
+		return -EINVAL;
+	}
+
+	data->is_alarm_pending = 0;
```

Comment:
please check and clear the status register of the RTC instead of caching the value, see `RTC_ISR_ALM_IS()`.

Reply:
I will remove the cached variable in the next update.